### DISCUSSION TITLE
Parsing command with lots of quotes `"` and `'`

### DISCUSSION THREAD

**Discussion by ghost** [awesomeWM/awesome#3929] [Q&A]
I have the following command to display CPU utilisation:

`top -bn1 | grep "Cpu(s)" | sed "s/.*, *\([0-9.]*\)%* id.*/\1/" | awk '{print 100 - $1"%"}'`

I can run it from terminal like so:

`bash -c "top -bn1 | grep Cpu\\(s\\) | sed --expression=\"s/.*, *\\([0-9.]*\\)%* id.*/\1/\""`

But when I put it inside awesome, I get an empty string:

`myCPUusage = awful.widget.watch('bash -c "top -bn1 | grep Cpu\\(s\\) | sed \"s/.*, *\\([0-9.]*\\)%* id.*/\1/\""', 1)`

Is there an alternative way of evaluating complex commands?

**Comment by actionless** [comment]
google://escape character in lua
google://multiline strings in lua

**Comment by ghost** [comment]
Thanks. It was solved by rewriting it in another way:

`command_to_print_CPU_usage='bash -c "top -bn1 | grep Cpu | cut -d , -f 4 | xargs | head -c -4 | cut -d. -f1 | awk \'{print 100 - $1\\"%\\"}\' "'`

**Comment by actionless** [comment]
also, that's offtopic from the point of your question, regarding escape characters, but i'd suggest to use short load average value instead of momentary utilization %:

```sh
bash -c 'echo scale=2\; $(cat /proc/loadavg | cut -d" " -f1) / $(nproc) | bc | sed "s/^\./0./g"'
```

or, to display it as %:

```sh
bash -c 'echo scale=2\; $(cat /proc/loadavg | cut -d" " -f1) / $(nproc) "*" 100 | bc | sed "s/^\./0./g"'
```

100% would mean full utilization of all cpus/cores, and >100% would mean that amount of processes is more than current cpus could handle, so some of them are queued

**Reply by ghost** [reply]
Thanks. It is a better command than mine, but I could not get it to work. The best I got is this, which produces an empty string:

`myCPUusage = awful.widget.watch('bash -c "echo scale=2\\; $\\(cat /proc/loadavg | cut -d\" \" -f1\\) / nproc | bc | sed \"s/^\\./0./g\""', 2)`

**Reply by actionless** [reply]
idk, try to experiment with escaping, for me it works:

![2024-06-28--1719572779_1452x66_scrot](https://github.com/awesomeWM/awesome/assets/1655669/c85b2a45-3150-43a9-baa3-a5606c7e94a9)

**Reply by ghost** [reply]
> for me it works:

Yes yes, from the terminal it works for me too. But I meant from inside awesomewm where it doens't work!

**Reply by actionless** [reply]
just try it little by little by adding one command after another, i know you can do it 💪

**Reply by actionless** [reply]
also, better use table instead of string for command: https://awesomewm.org/doc/api/classes/awful.widget.watch.html#awful.widget.watch:new

**Reply by actionless** [reply]
btw i just noticed that scale= parameter in bc affecting not just display of the value, but also computation precision, so the better way do it like that:

```
bash -c 'printf "%2.2f%%\n" $(echo scale=10\; $(cat /proc/loadavg | cut -d" " -f1) / $(nproc) "*" 100 | bc )'
```
